SPEC.md

```diff
@@ -935,3871 +1811,7403 @@ In operations on mismatched numeric types (e.g. `Int` + `Float`), the `Int` type
+
+Provides input-specific hints in the form of an object. Each key within this hint should refer to an actual input defined for the current task. A key may also refer to a specific member of a struct/object input.
+
+<details>
+<summary>
+Example: input_hint_task.wdl
 
 ```wdl
-task test {
+version 1.1
+
+struct Person {
+  String name
+  File? cv
+}
+
+task input_hint {
   input {
-    String? val
+    Person person
   }
 
   command <<<
-    python script.py --val=~{val}
+  if ~{defined(person.cv)}; then
+    grep "WDL" ~{person.cv}
+  fi
   >>>
+  
+  output {
+    Array[String] experience = read_lines(stdout())
+  }
+
+  runtime {
+    inputs: object {
+      person: object {
+        cv: object {
+          localizationOptional: true
+        }
+      }
+    }
+  }
 }
 ```
+</summary>
+<p>
+Example input:
 
-Since `val` is optional, the command can be instantiated in two ways:
-
-```
-python script.py --val=foobar
+```json
+{
+  "input_hint.person": {
+    "name": "Joe"
+  }
+}
 ```
 
-Or
+Example output:
 
+```json
+{}
 ```
-python script.py --val=
-```
+</p>
+</details>
+
+Reserved input-specific attributes:
+
+* `inputs.<key>.localizationOptional`: Tells the execution engine that a specific `File` input does not need to be localized for this task.
+
+##### `outputs`
+
+* Allowed type: `object`
+
+Provides outputs specific hints in the form of a hints object. Each key within this hint should refer to an actual output defined for the current task. A key may also refer to a specific member of a struct/object input.
+
+#### Conventions and Best Practices
+
+In order to encourage interoperable workflows, WDL authors and execution engine implementors should view hints strictly as an optimization that can be made for a specific task at runtime; hints should not be interpreted as requirements for that task. By following this principle, we can guarantee that a workflow is runnable on all platforms assuming the `runtime` section has the required parameters, regardless of whether it contains any additional hints.
+
+The following guidelines should be followed when using hints:
+
+* A hint should never be required.
+* Less is more. Before adding a new hint key, ask yourself "do I really need another hint?", or "is there a better way to specify the behavior I require". If there is, then adding a hint is likely not the right way to achieve your aims.
+* Complexity is killer. By allowing any arbitrary keys, it is possible that the `runtime` section can get quite unruly and complex. This should be discouraged, and instead a pattern of simplicity should be stressed.
+* Sharing is caring. People tend to look for similar behavior between different execution engines. It is strongly encouraged that execution engines agree on common names and accepted values for hints that describe common usage patterns. A good example of hints that have conventions attached to them is cloud provider specific details:
+    ```wdl
+    task foo {
+      .... 
+      runtime {
```

Comment:
was the decision for our last call to rename `hints` to `runtime` and make them functionally equivalent? we added the new `requirements` section

Reply:
In 1.2 we will have `requirements` and `hints`. `runtime` will be deprecated but still supported until `2.0`. The use of `runtime` is orthogonal to the use of `requirements` + `hints` - i.e. you can do one or the other but not mix them.